sfdx-source/apex-common/main/classes/fflib_Criteria.cls

```diff
@@ -0,0 +1,654 @@
+/**
+ * Copyright (c), FinancialForce.com, inc
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ *   are permitted provided that the following conditions are met:
+ *
+ * - Redistributions of source code must retain the above copyright notice,
+ *      this list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright notice,
+ *      this list of conditions and the following disclaimer in the documentation
+ *      and/or other materials provided with the distribution.
+ * - Neither the name of the FinancialForce.com, inc nor the names of its contributors
+ *      may be used to endorse or promote products derived from this software without
+ *      specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ *  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ *  OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ *  THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ *  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ *  OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ *  OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**/
+public virtual with sharing class fflib_Criteria
+		implements Evaluator
+{
+	// Add brace when generated to SQL
+	private Boolean embraced = false;
+
+	// List to hold all evaluators
+	private List<Evaluator> evaluators;
+
+	// The type of comparator NONE | AND | OR
+	private String type;
+
+
+	/**
+	 * populate private variables with default settings.
+	 */
+	public fflib_Criteria()
+	{
+		this.evaluators = new List<Evaluator>();
+		this.type = 'AND';
+	}
+
+	/**
+	 * Changes the default comparator for each criteria to OR
```

Comment:
Is it more accurate to say “changes the default comparator for *the next* criteria...”?

Reply:
@SWhalen-gh It will change the default comparator for the entire criteria not only the next. Or you need to combine an AND and OR in one query it should be encapsulated between braces. e.g.:
`something AND (x OR y)`. The `x OR y` is another instance of fflib_Criteria, added via the method `addOrCriteria`

Reply:
That explanation clarifies things.  But these blocks of code are dangerously similar:

.equalTo(...)
.equalTo(...)
.orCriteria(...)   
.equalTo();

//compared to:
.equalTo(...)
.equalTo(...)
.addOrCriteria(...)
.equalTo(...)

It would be clearer if the names of the methods were more distinct, like “setDefaultComparator()” instead of “orCriteria()” ...Or, have “orCriteria()” throw an error if it is not the first statement in the block.

Reply:
It would look like:
```
new fflib_Criteria()
    .equalTo(...)
    .equalTo(...)
    .addOrCriteria(
         new fflib_Criteria()
             .equalTo(..)
             .equalTo(..)
   )
```
Would result in something like: x AND y AND  (k OR m)

```
new fflib_Criteria()
    .orCriteria()
    .equalTo(...)
    .equalTo(...)
    .addAndCriteria(
         new fflib_Criteria()
             .equalTo(..)
             .equalTo(..)
   )
```
Would result in something like: x OR y OR (k AND m)

Reply:
It is also not the intention to have very large and complex conditions. As you can combine the conditions as shown in the example.

It is however a good idea to make sure the orCriteria method should be called first in the statement block.